### ISSUE TITLE
Provide more feedback / add verbosity options

### ISSUE THREAD

**Issue by dougthor42** [sixty-north/cosmic-ray#219]
# Summary
It would be super awesome if some verbosity were added to the two main commands, `init` and `exec`.

Here are some suggestions.

## init
During `cosmic-ray init`, it would be nice to see how many mutations will tested and where they are. 3 levels of additional verbosity could be defined:

#### 0. No Verbosity
```
$ cosmic-ray init --baseline=2 test_session my_pkg -- my_pkg_tests
Created 2368 mutations to Test.
```
#### 1. Summary of how many of each mutation are to be tested

`-v`, `--verbosity=1`

```
$ cosmic-ray init --baseline=2 test_session my_pkg -- my_pkg_tests -v
Created 2368 mutations to test.
   1137 "mutate_unary_operator" mutations
   792 "number_replacer" mutations
   211 "mutate_comparison_operator" mutations
   196 "mutate_binary_operator" mutations
   22 "replace_true_false" mutations
   6 "replace_or_with_and" mutations
   4 "add_not" mutations
```

#### 2. Start/End Times and List which package/module is currently being processed.

`-vv`, `--verbosity=2`

Also display the info from level 1 verbosity. Note that the mutation count list is sorted in descending order.

```
$ cosmic-ray init --baseline=2 test_session my_pkg -- my_pkg_tests -vv
Start: 2017-06-21 09:20:19
Creating mutations for my_pkg.main... Done
Creating mutations for my_pkg.module1... Done
Creating mutations for my_pkg.module2... Done
Created 2368 mutations to Test.
   1137 "mutate_unary_operator" mutations
   792 "number_replacer" mutations
   211 "mutate_comparison_operator" mutations
   196 "mutate_binary_operator" mutations
   22 "replace_true_false" mutations
   6 "replace_or_with_and" mutations
   4 "add_not" mutations
Finish 2017-06-21 09:22:36
```

#### Breakdown of Mutations by Package/Module

`-vvv`, `--verbosity=3`

Also all information from previous levels.

> Note: For this example, I've given up on trying to make the counts add up correctly. Sorry!

```
$ cosmic-ray init --baseline=2 test_session my_pkg -- my_pkg_tests -vvv
Start: 2017-06-21 09:20:19
Creating mutations for my_pkg.main... Done
Creating mutations for my_pkg.module1... Done
Creating mutations for my_pkg.module2... Done
my_pkg.main
   1234 total mutations
   637 "number_replacer" mutations
   236 "mutate_unary_operator" mutations
   ... etc
my_pkg.module1
   698 total mutations
   316 "mutate_unary_operator" mutations
   113 "number_replacer" mutations
   ... etc
my_pkg.module2
   436 total mutations
   637 "mutate_unary_operator" mutations
   236 "number_replacer" mutations
   ... etc
Created 2368 mutations to Test.
   1137 "mutate_unary_operator" mutations
   792 "number_replacer" mutations
   211 "mutate_comparison_operator" mutations
   196 "mutate_binary_operator" mutations
   22 "replace_true_false" mutations
   6 "replace_or_with_and" mutations
   4 "add_not" mutations
Finish 2017-06-21 09:22:36
```

## exec
During `cosmic-ray exec`, it would be nice to see which mutation it's currently on.

Without any verbosity added, it could just be a the finish and elapsed time and single line that rewrites itself. You could also use the [tqdm](https://pypi.python.org/pypi/tqdm) or [progressbar2](http://progressbar-2.readthedocs.io/en/latest/) packages.

```
$ cosmic-ray exec test_session
Mutation 1/2368...

(a little while later...)
$ cosmic-ray exec test_session
Mutation 79/2368...

(when done)
$ cosmic-ray exec test_session
Mutation 2368/2368...
Mutation testing completed at 2017-06-21 09:57:11
Elapsed time: 4 hours, 39 minutes, 24 seconds
```

With verbosity, it could list out the timestamp, mutation number, and elapsed time:

```
$ cosmic-ray exec test_session -v
2017-06-21T09:37:45 Mutation 1/2368...    2m39s
2017-06-21T09:40:24 Mutation 2/2368...    2m26s
2017-06-21T09:42:50 Mutation 3/2368...
```

**Comment by abingham** [comment]
I definitely agree that our verbosity support could be improved. 

There's already a global "verbosity" setting that will effect all of the subcommands. It works by changing the logging level used by the `logging` package, and currently it's quite simple: by default we only log WARNING or higher (I think), and in verbose mode we log INFO and higher (almost everything). This could certainly be made more sophisticated. My instinct, however, is to stick with a single verbosity setting rather than adding specific new flags for individual commands. In any event, I think docopt would complain if we tried to have more than one flag named verbosity.

So I'd be happy with changing the current, simple verbosity control to something with multiple levels. But I like the idea of verbosity being tied to logging levels. It's a simple, explainable, and understandable model that will work well across plugins, etc. I wouldn't be opposed to adding some custom logging levels if that turns out to be necessary, but I'd need to be convinced that such complexity is really helpful.

I'll handle your individual suggestions in separate comments.

**Comment by abingham** [comment]
> During cosmic-ray init, it would be nice to see how many mutations will tested and where they are. 

While I agree that this is a potentially interesting piece of information, I'm not sure that `init` is the right place for printing it. `init` is a very heavy-handed command in that it wipes out any previous testing results. So if users are generally interested in knowing how many mutations will be performed, they'll want to be able to do that without re-running `init`, i.e. we need a separate command that can provide that piece of information. Looked at this way, printing the number of mutations as part of `init` is redundant; we are obligated to provide another means of getting the same information.

Ultimately, I would like to use CR's reporting system for this kind of stuff. The plan (such as it is) is to have a command which prints a JSON representation of the current session. This JSON can be piped into other commands which parse it and produce whatever information is desired (e.g. number of mutations, mutations collated with packages, survival rates, etc.) This will cleanly separate the CR core from all the kinds of reports users might want. I've got a working version of this in [the `report-rework` branch](https://github.com/sixty-north/cosmic-ray/tree/report-rework); I just need to find time to clean it up and get it into master.

**Comment by abingham** [comment]
> 1. Summary of how many of each mutation are to be tested . . .

For the other proposed verbosity levels of init, I feel pretty strongly that a separate tool based on a generic JSON dump is the way to go. If you look at [`commands/format.py` on the `report-rework` branch](https://github.com/sixty-north/cosmic-ray/blob/report-rework/cosmic_ray/commands/format.py) I think you'll get a sense of what I mean.

**Comment by abingham** [comment]
> During cosmic-ray exec, it would be nice to see which mutation it's currently on.

I'm pretty sure this is not a well-defined concept. In a distributed configuration, `exec` could be running an arbitrary number of mutations in parallel, so there's generally no notion of "current mutation".

You can, however, determine how many mutations have been tested so far, and I think this is spiritually the same information. When I want to track progress right now I just run `cosmic-ray report`, and it tells me that information. Following the theme I started earlier, my inclination here is to provide this sort of progress data via a separate reporting tool, not as part of `exec`.

**Comment by abingham** [comment]
So I hope I'm not coming across as too pessimistic. I definitely want to encourage the kind of work you're suggesting, but I need to constrain the implementation to my current vision lest the implementation sprawls.

I'd like to get the `report-rework` branch into main soon. It's really just a question of me finding the time. Would you be willing to try implementing your ideas against that branch using the "standalone reporting tools" approach I'm advocating?

**Comment by dougthor42** [comment]
**First, a caveat:** I am a new user, having just heard about this package by watching the PyCon 2017 video [Next Level Testing](http://pyvideo.org/pycon-us-2017/next-level-testing.html) by James Saryerwinnie of the  AWS "SDK and Tools" Team. I jumped over to your docs and dove right in, testing `cosmic-ray` on some of my more well-tested and well-covered projects.

Now back to our regularly scheduled discussion.

----

**TL;DR:**

+ The `--verbose` option that I overlooked in the docs seems to solve most of the comments relating to logging/progress/status during `exec`
+ I'm 100% for a standalone reporting tool for the requests I'm proposing. My only wish is that `init` (optionally?) calls that reporting tool after running.

----

> There's already a global "verbosity" setting

Hmm... I didn't notice that when reading through the docs. I'll check again.  

....... Yup, there it is, hidden in a couple sections and also smacking me right in the face with `cosmic-ray help`. My bad!

> stick with a single verbosity setting rather than adding specific new flags for individual commands. 

Understandable. I was using the [`green`](https://github.com/CleanCut/green) project as a reference, but they don't have multiple command verbs like `cosmic-ray` does.

> I think docopt would complain if we tried to have more than one flag named verbosity.

Actually, it [handles it quite well](https://stackoverflow.com/a/18487687/1354930), at least for short-form args. I'm not sure how well it's documented though.

----

> I'm not sure that `init` is the right place for printing [the number of mutations]. ... they'll want to be able to do that without re-running `init`. ... printing the number of mutations as part of `init` is redundant.

I agree that a separate command (`cosmic-ray stats my_session` or something) is the way to go - it offers the most flexibility. All I ask is that the `init` command (optionally?) runs `stats` automatically upon completion.

I'm looking forward to what you're adding in the `report-rework` branch.

----

>> 1. Summary of how many of each mutation are to be tested . . .
>
> For the other proposed verbosity levels of `init`, I feel pretty strongly that a separate tool based on a generic JSON dump is the way to go.

Totally acceptable. As I mentioned above, it could be tacked on to the end of `init` very easily. Heck it wouldn't even have to be part of your source: the docs could just tell users to chain the commands together (though I admit, I'd prefer it to just be a flag on `init` 😄)

```
$ cosmic-ray init my_session module -- tests && cosmic-ray stats my_session
```

----

> exec could be running an arbitrary number of mutations in parallel, so there's generally no notion of "current mutation". You can, however, determine how many mutations have been tested so far, and I think this is spiritually the same information.

Yes, that's a very good point. I'd be perfectly fine with seeing `x/total` flash by my screen. Again, using `--verbose` solves this in a roundabout fashion. It fills up the terminal, which isn't preferred, but is not a dealbreaker.

> When I want to track progress right now I just run `cosmic-ray report`, and it tells me that information.

So you're saying that you can run `cosmic-ray report` on a session that's currently running? Yup, totally can. That's awesome. That info should probably be added to the docs. (I'll open a PR).

----

> So I hope I'm not coming across as too pessimistic.

Not at all 😁.

> I need to constrain the implementation to my current vision lest the implementation sprawls.

Yup, feature-creep kills many a project

> Would you be willing to try implementing your ideas against that branch using the "standalone reporting tools" approach I'm advocating?

Of course. My naïve initial thought is that it shouldn't be too hack `init` to call `report` upon completion or to have `exec` call `report` periodically. But I've also not looked at the codebase so... /shrug.

**Comment by abingham** [comment]
@dougthor42 I've merged the `report-rework` branch into main. There are a number of internal changes, but the main user-visible change is that `cosmic-ray report` has been replaced by `cosmic-ray dump`. `dump` prints  a JSON version of the session to stdout, ready for consumption by other tools, of which we provide a few: `cr-rate` which simply prints the session's survival rate (we need this for automated testing) and `cr-report` which does what `report` used to do.

**Comment by dougthor42** [comment]
I'll see what I can whip up. I can't guarantee how much time I'll be able to spend on it, but as I said earlier, it should be fairly easy to parse that JSON.

Heck, there's probably already a library that calculates stats on JSON arrays...

**Comment by boxed** [comment]
@dougthor42  Sorry to spam here, but I couldn't find any way to contact you directly. You could give my mutation tester mutmut (https://github.com/boxed/mutmut/) a try.

**Comment by alexvicegrab** [comment]
Documentation and the CLI are not aligned currently. Would be great if we can update documentation on verbosity levels.

Thanks for this excellent tool! :)

**Comment by disouzam** [comment]
I'm late to the game but I liked the idea proposed here by @dougthor42 ! I started using Cosmic Ray last year and I'm returning to use it this year and would like to contribute somehow (there is a new issue and new pull request in another direction - #598 and #599 ) but this one looks interesting to me. 

The project I'm trying to add mutation testing, another OSS software - [pycalphad](https://github.com/pycalphad/pycalphad), has 26k mutants and I'm having some difficulties to deal with all of them. Verbosity for exec is not enough to get current status and I'm doing a mix of using cr-report and using database directly using custom SQL scripts.
